File: Python/Gemini 1.5 Flash/Observer/round1/Subject.py
class Subject:
    """Represents the car, which is the subject of observation."""

    def __init__(self):
        self._observers = []

    def attach(self, observer):
        """Attach an observer to the car."""
        self._observers.append(observer)

    def detach(self, observer):
        """Detach an observer from the car."""
        self._observers.remove(observer)

    def notify(self, event):
        """Notify all observers about the event."""
        for observer in self._observers:
            observer.update(event)

File: app/src/test/java/sqa_project/GeminiPro/observer/round1/Subject.java
package sqa_project.GeminiPro.observer.round1;

import java.util.ArrayList;
import java.util.List;


public class Subject {
    private final List<CarObserver> observers = new ArrayList<>();
    private String featureUpdate;

    // Method to add an observer
    public void addObserver(CarObserver observer) {
        observers.add(observer);
    }

    // Method to remove an observer
    public void removeObserver(CarObserver observer) {
        observers.remove(observer);
    }

    // Method to set feature update and notify all observers
    public void setFeatureUpdate(String featureUpdate) {
        this.featureUpdate = featureUpdate;
        notifyAllObservers();
    }

    // Method to notify all observers of a feature update
    private void notifyAllObservers() {
        for (CarObserver observer : observers) {
            observer.update(featureUpdate);
        }
    }
}

